File: src/main/java/cc/mirukuneko/realtrainmodrenewed/network/CarScriptDataPayload.java
package cc.mirukuneko.realtrainmodrenewed.network;

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed;
import cc.mirukuneko.realtrainmodrenewed.entity.CarEntity;
import io.netty.buffer.ByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;
import net.minecraft.world.entity.Entity;
import net.neoforged.neoforge.network.handling.IPayloadContext;

/**
 * クライアント→サーバの scriptData(DataMap) 同期。
 *
 * <p>SuperRailBuilder3 等は render(クライアント)スクリプトで GUI 入力を受け取り、設置点や
 * isBuilding 等を DataMap へ書く。サーバの onUpdate がそれを読んで実際にレールを敷くため、
 * クライアントの書き込みをサーバの同一エンティティへ届ける必要がある(RTMUは従来サーバ→
 * クライアント同期しか無かった)。</p>
 */
public record CarScriptDataPayload(int entityId, String key, String value) implements CustomPacketPayload {
    public static final Type<CarScriptDataPayload> TYPE = new Type<>(
        Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "car_script_data")
    );

    public static final StreamCodec<ByteBuf, CarScriptDataPayload> STREAM_CODEC = StreamCodec.composite(
        ByteBufCodecs.INT, CarScriptDataPayload::entityId,
        ByteBufCodecs.STRING_UTF8, CarScriptDataPayload::key,
        ByteBufCodecs.STRING_UTF8, CarScriptDataPayload::value,
        CarScriptDataPayload::new
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }

    public static void handleOnServer(CarScriptDataPayload payload, IPayloadContext context) {
        context.enqueueWork(() -> {
            if (context.player() == null) {
                return;
            }
            Entity entity = context.player().level().getEntity(payload.entityId());
            if (entity instanceof CarEntity car) {
                car.setScriptDataValue(payload.key(), payload.value());
            }
        });
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/network/CarScriptDataPayload.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.network

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import cc.mirukuneko.realtrainmodrenewed.entity.CarEntity
import io.netty.buffer.ByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import net.minecraft.server.level.ServerPlayer
import net.neoforged.neoforge.network.PacketDistributor
import net.neoforged.neoforge.network.handling.IPayloadContext

/** Client-to-server scriptData/DataMap sync for script-driven CarEntity state. */
@JvmRecord
data class CarScriptDataPayload(
    val entityId: Int,
    val key: String,
    val value: String,
    val flags: Int,
) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

    companion object {
        @JvmField
        val TYPE: CustomPacketPayload.Type<CarScriptDataPayload> = CustomPacketPayload.Type(
            Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "car_script_data")
        )

        @JvmField
        val STREAM_CODEC: StreamCodec<ByteBuf, CarScriptDataPayload> = StreamCodec.composite(
            ByteBufCodecs.INT,
            { payload -> payload.entityId },
            ByteBufCodecs.STRING_UTF8,
            { payload -> payload.key },
            ByteBufCodecs.STRING_UTF8,
            { payload -> payload.value },
            ByteBufCodecs.INT,
            { payload -> payload.flags },
            ::CarScriptDataPayload,
        )

        @JvmStatic
        fun handleOnServer(payload: CarScriptDataPayload, context: IPayloadContext) {
            context.enqueueWork {
                val player = context.player() as? ServerPlayer ?: return@enqueueWork
                val car = player.level().getEntity(payload.entityId) as? CarEntity ?: return@enqueueWork
                car.applyClientScriptData(player, payload.key, payload.value, payload.flags)
                if (payload.key.length <= 64) {
                    PacketDistributor.sendToPlayer(
                        player,
                        CarScriptDataSyncPayload(
                            car.id,
                            mapOf(payload.key to car.getScriptDataValue(payload.key)),
                        ),
                    )
                }
            }
        }
    }
}
